src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                var typeError = SyntaxFactory.FunctionPointerParameterOrReturnType(default, identifierError);
+                typesError.Add(typeError);
+                var greaterThanTokenError = SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken);
+                greaterThanTokenError = WithAdditionalDiagnostics(greaterThanTokenError, GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
+
+                var funcPtr = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanTokenError, typesError, greaterThanTokenError);
+                _pool.Free(typesError);
+                return funcPtr;
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
+                        {
+                            modifiers.Add(EatToken());
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
```

Comment:
this feels like a bug right?  or are you really supposed to pass null here and ignore any result?  are all skipped tokens properly added to the `types` collection passed below?

Reply:
Yeah, this is really how it's supposed to work. See [http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,f32aaddddf4a6d5b](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,f32aaddddf4a6d5b) as the pre-existing example I copied this from. It definitely does feel like the type of pattern that would not have been written if NRT was a thing since day 1.

---
In reply to: [349284251](https://github.com/dotnet/roslyn/pull/39879#discussion_r349284251) [](ancestors = 349284251)